Datapath mode

mkalcok opened the pull request:
Work in progress to allow MicroOVN to operate in datapath-only mode with external ovn-central services.

please review commit-by-commit

mkalcok commented on microovn/ovn/environment/environment.go:
```diff
@@ -220,3 +221,42 @@ func CleanupPaths() error {
 
 	return errors.Join(errs...)
 }
+
+// UpdateOvnListenConfig configures the OVN NB and SB databases to listen on the appropriate ports.
+func UpdateOvnListenConfig(ctx context.Context, s state.State) error {
```
To me it seemed appropriate to have these function in a package that manipulates environment to alter OVN's runtime behavior, but I also see your point that this file was previously dedicated more to the environment variables and cli argument values.

On the other hand, i kinda like that the "cmd" package exposes only generic wrappers for running general OVN cli commands.

Perhaps a new package would be a solution here?

fnordahl replied:
Yes new package also works for me. Composition FTW!

mkalcok commented on microovn/cmd/microovn/init.go:
```diff
@@ -115,6 +115,16 @@ func (c *cmdInit) Run(_ *cobra.Command, _ []string) error {
 		optionalConfig := make(map[string]string)
+
+		// TODO: Perhaps we need to change this question to get comma separated values
+		selectedServices, err := c.common.asker.AskChoice("Please select services you would like to enable on this node (central/chassis/auto) [default=auto]: ", []string{"central", "chassis", "auto"}, "auto")
```
yeah, I was thinking about this and I was wondering if there's a practical use-case for running switch service only. Should we perhaps consolidate roles to only `central` and `chassis`?

Note: There are pieces of code dating back to original implementation that treat switch/chassis service role interchangeably, those should be probably fixed if we keep it. (It shouldn't be too much work)

fnordahl replied:
There is a practical use case for running switch service only.

The MicroOVN snap is our preferred distribution channel for Open vSwitch and OVN, when in use it is the sole source of Open vSwitch on the sytem.

When using the userspace data path (DPDK, OVS-DOCA), or kernel mode hardware offlload a Open vSwitch bridge will be responsible for all networking for a host with no oob management. And yes there are customers with these requirements.

https://bugs.launchpad.net/microovn/+bug/2045490
https://github.com/canonical/netplan/pull/549
https://docs.openstack.org/project-deploy-guide/charm-deployment-guide/latest/install-maas.html#create-ovs-bridge

fnordahl reviewed (changes requested):
Thank you for your swift work on this!

This is not a full review, although I've read through the changes to the processing of internal vs. external source of SB DB connection string, and the approach seems reasonable to me.

I have an ask for namespacing the config command, but I only expect us to do this in the CLI level for now, see in-line comment.

mkalcok commented on microovn/node/node.go:
```diff
@@ -40,13 +41,16 @@ func DisableService(ctx context.Context, s state.State, service types.SrvName) e
 			return err
 		}
+		// TODO: We should probably have some --i-really-mean-it flag
+		// from the client request before allowing this
```
anti-foot-gun measures were applied by requiring `--allow-disable-last-central` flag to disable last central node.

fnordahl commented on microovn/ovn/environment/environment.go:
```diff
@@ -17,7 +17,6 @@ import (
 	"github.com/canonical/lxd/shared/logger"
 	"github.com/canonical/microcluster/v2/cluster"
 	"github.com/canonical/microcluster/v2/state"
-
```
Is removal of this blank line intentional?

It provided logical separation of imports, which I think we'd like to keep.

mkalcok replied:
Thanks for catching this. It wasn't intentional.

fnordahl replied:
nit: thanks for putting it back, it appears to be removed in commit 2 and put back in commit 5, so if you have an opportunity to move it to commit 2 that would give a cleaner git history.

fnordahl commented on microovn/ovn/join.go:
```diff
@@ -48,9 +48,15 @@ func Join(ctx context.Context, s state.State, initConfig map[string]string) erro
+	}
 	enableServices := requestedServices{
-		Central: srvCentral < 3,
+		Central: !externalOvnCentral && srvCentral < 3,
```
Thank you for adding this, it's a nice compromise to expedite charm development in canonical/microovn-operator#21

fnordahl reviewed (commented):
Thank you for your swift and thorough work on this, it looks good to me, pending some updates to commit messages, newline changes and docs.

I love the test coverage!

I also like the approach of configuring the IPs separately, which both makes the interface simpler for a human to understand (they do not need to know the intricate details of OVSDB connection strings to use it), and secure by default by enforcing ssl.

There are a few commits which need updating of subject (`WIP` prefix) and content (naming changed during review)to match the actual state of implementation, but I assume you will update these when you add docs.

mkalcok commented:
Thank you for the review.
Yes the `WIP` tags in commits mark those that still need doc update, just so I won't forget anything :laughing:

I'll double-check for naming changes, I thought that I covered those already, but looks like I missed some spots.

mkalcok commented:
The documentation was added and the comments from previous round of reviews were addressed. Regarding the docs it was mainly additions, I changed couple of other things I encountered. In the `service-control` docs I also changed:
* emphasis that the `snap` is not recommended way to adjust services
* Titles that used gerunds, because our documentation guide advises against them

This should be now fully prepared for review and merge

mkalcok commented:
Interesting, a test (ovn_ovs_cli_ipv6) that was [passing fine](https://github.com/canonical/microovn/actions/runs/16319557888/job/46093950890) before, seems to be stuck now after fixing a typo in docs :thinking:
I'll investigate further.

fnordahl commented on docs/how-to/datapath-only-mode.rst:
```diff
@@ -0,0 +1,112 @@
+
+This is a mode of operation in which MicroOVN cluster runs only ``switch`` and
+``chassis`` service, with ``central`` service being provided externally by another
+MicroOVN cluster or by some other OVN deployment.
```
It is pertinent to put emphasis on the main use case being migration / transition.

In the current wording a unsuspecting user might mistake this feature to OVN interconnect or similar where two deployments actually can work together, which is not the case for this particular feature.

fnordahl commented on docs/how-to/datapath-only-mode.rst:
```diff
@@ -0,0 +1,112 @@
+
+.. code-block::
+
+   $ sudo microovn init
```
No terminal prefix in code-blocks, make two code-blocks instead, one with the command, another with the output/interaction.

fnordahl commented on docs/how-to/datapath-only-mode.rst:
```diff
@@ -0,0 +1,112 @@
+
+.. code-block::
+
+   $ sudo microovn init
```
No terminal prefix in code-blocks, make two code-blocks instead, one with the command, another with the output/interaction.

fnordahl commented on docs/how-to/datapath-only-mode.rst:
```diff
@@ -0,0 +1,112 @@
+
+.. code-block::
+
+   $ sudo microovn.ovn-sbctl show
```
No terminal prefix in code-blocks, make two code-blocks instead, one with the command, another with the output/interaction.
